### ISSUE TITLE
request: more visibility into JSON parsing errors

### ISSUE THREAD

**Issue by Phrohdoh** [hediet/vscode-debug-visualizer#78]
With this (completely valid and schema-compliant (`AstTreeVisualizationData`), AFAICT) JSON

```json
{"kind":{"ast":true,"tree":true,"text":true},"root":{"items":[{"text":"exploding-barrel:"}],"children":[{"items":[{"text":"Tooltip:"}],"children":[],"span":{"start":18,"length":13},"segment":"todo"},{"items":[{"text":"Name: barrels"}],"children":[],"span":{"start":31,"length":22},"segment":"todo"},{"items":[{"text":"Health:"}],"children":[],"span":{"start":53,"length":12},"segment":"todo"},{"items":[{"text":"HP: 5"}],"children":[],"span":{"start":65,"length":14},"segment":"todo"},{"items":[{"text":"Explodes:"}],"children":[],"span":{"start":79,"length":14},"segment":"todo"},{"items":[{"text":"Weapon: large-barrel-explode"}],"children":[],"span":{"start":93,"length":37},"segment":"todo"},{"items":[{"text":"MapEditorData:"}],"children":[],"span":{"start":130,"length":19},"segment":"todo"},{"items":[{"text":"Categories: props, dangerous-props"}],"children":[],"span":{"start":149,"length":43},"segment":"todo"}],"span":{"start":0,"length":193},"segment":"doc-root"},"text":"exploding-barrel:\n    Tooltip:\n        Name: barrels\n    Health:\n        HP: 5\n    Explodes:\n        Weapon: large-barrel-explode\n    MapEditorData:\n        Categories: props, dangerous-props\n"}
```

I am getting the following parse error:

<img width="644" alt="Screen Shot 2020-08-30 at 8 36 08 AM" src="https://user-images.githubusercontent.com/4861023/91660504-f9cccb00-ea9b-11ea-9432-06532f715b97.png">

Given my JSON is valid, I'm just _stuck_.
I don't know what to change (and why) to give the visualizer what it wants.

Any insight and ideas on improving this greatly appreciated!

**Comment by hediet** [comment]
It looks like it tries to parse it as string (`"{"`) and then encounters the `k`.

What is the difference to https://github.com/hediet/vscode-debug-visualizer/pull/75?

**Comment by hediet** [comment]
[Also, to test the schema, you can use the playground](https://hediet.github.io/visualization/?darkTheme=1&state=XQAAAAKtBAAAAAAAAABAqQlHM6T2WdDo2qjJ7RXQXlsilxQOyi4j_UI9WR_L7GT_ktUMXhyGbaWGiznI20OA5fXRzcKPMqoz40cNysVeqvQChYQIuElCNqOxPSZZmRA8x8sh91aCeDFOqwdIVuQUInTKG_ChK7NwNSOMKXcBTS6LHhTf12ewnh9AjX4Tfi43JJuzD9mBuXOJAY5G-KBe1QNn0sfemMG4wRmpv6DKcvVNMU1UvW4iNYnVZZ5MxbeM1q0qtUlE15RE0NjOAo6xxq6o_tRvWOVUp0sfmqZy2-EFvub9EsA4lrwDSMnVXwXP7LcQcfC6kOA0Wkdw6gxa-vEPEyW1uhCjvwYqsmHonI8YSDt0_oEBxlIJMuKYwDArMozp_WPjNpTrVUClRBsXUe8asJ7gm_DpqK3S72iRWDGhPMZ0Gnl4GP2MS6aWVqMs1hXeB3YIGfBB1_2WlSYO9smNQQud0oTr_4mOdUA).

**Comment by hediet** [comment]
My guess is that lldb shortens long strings. The debug visualizer extension is trying out various ways how to parse the string (it tries to remove any surrounding quotation mark), but only reports the last one.
So there might be a different json parse error. Can to copy the entire text of the error message and paste it here?

**Comment by Phrohdoh** [comment]
Yeah, it looks like lldb truncates 🤦.

```
Could not parse evaluation result as JSON:

Unexpected token k in JSON at position 3

Evaluation result was:
"{"kind":{"ast":true,"tree":true,"text":true},"root":{"items":[{"text":"exploding-barrel:"}],"children":[{"items":[{"text":"Tooltip:"}],"children":[],"span":{"start":18,"length":13},"segment":"todo"},{"items":[{"text":"Name: barrels"}],"children":[],"span":{"start":31,"length":22},"segment":"todo"},{"items":[{"text":"Health:"}],"children":[],"span":{"start":53,"length":12},"segment":"todo"},{"items":[{"text":"HP: 5"}],"children":[],"span":{"start":65,"length":14},"segment":"todo"},{"items":[{"text":"Explodes:"}],"children":[],"span":{"start":79,"length":14},"segment":"todo"},{"items":[{"text":"Weapon: large-barrel-explode"}],"children":[],"span":{"start":93,"length":37},"segment":"todo"},{"items":[{"text":"MapEditorData:"}],"children":[],"span":{"start":130,"length":19},"segment":"todo"},{"items":[{"text":"Categories: props, dangerous-props"}],"children":[],"span":{"start":149,"length":43},"segment":"todo"}],"span":{"start":0,"length":193},"segment":"doc-root"},"text":"exploding-barrel:..."

Used debug adapter: lldb
```

**Comment by Phrohdoh** [comment]
https://stackoverflow.com/a/41600000 may be what we want in this case.

**Comment by hediet** [comment]
Or this one: https://stackoverflow.com/questions/31402092/print-long-string-in-xcode-6-debugging-console

**Comment by hediet** [comment]
https://github.com/vadimcn/vscode-lldb/blob/master/MANUAL.md#launching-a-new-process

I think `initCommands` can be used to configure these things.

**Comment by Phrohdoh** [comment]
Setting `target.max-string-summary-length` to `10000` (via `initCommands`) seems to make no difference, the string is still truncated (with `...` appended) at the exact same position.

**Comment by hediet** [comment]
Your string is truncated after exactly 1000 chars. Maybe you can reverse search a setting that has that value.

**Comment by hediet** [comment]
[This article](https://medium.com/@nixhax/pretty-print-dictionaries-as-json-in-lldb-44913964ff2a) also uses `target.max-string-summary-length`.
Is the watch windows also truncating the string? What about the debug console?

**Comment by Phrohdoh** [comment]
AFAIK the watch viewlet and debug console are driven by the debug adapter.
Yes, both of them display the truncated value.

I do not have any lldb settings with the value `1000` (confirmed by running `settings show` and searching over the result).

**Comment by hediet** [comment]
> AFAIK the watch viewlet and debug console are driven by the debug adapter.

Yes, this extension is driven too by the debug adapter. However, the watch viewlet uses the `watch` context, the debug console the `repl` context, and this extension by default also the `repl` context (that you overwrote with `watch`).

**Comment by hediet** [comment]
> I do not have any lldb settings with the value 1000 (confirmed by running settings show and searching over the result).

This is unfortunate. If you make it work in the repl context with long strings (maybe there is some special formatting argument), I could extend the json parsing logic to deal with lldb outputs.

If lldb has no way to output long strings, this extension cannot do much.

**Comment by Phrohdoh** [comment]
This is indeed unfortunate.
Thank you for your help and time!

**Comment by hediet** [comment]
I think your main point is still valid. The error reporting is still not good.
